File: Data Structures & Algorithms/binary-tree-diameter/submission-8.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDiameter = 0;
    int height(TreeNode* node){
        if(node == nullptr) return 0;
        int heightLeft = height(node->left);
        int heightRight = height(node->right);
        if(heightLeft + heightRight > maxDiameter){
            maxDiameter = heightLeft + heightRight;
        }
        return 1 + max(heightLeft,heightRight);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == nullptr) return 0;
        height(root);
        return maxDiameter;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-8.py
class Solution:
    '''
    left = 1, right = 6 ==> 7*1 = 7 ==> l < r ==> left ++, max = 7
    left = 7, right = 6 ==> 6*6 =36 ==> r < l ==> right -- , max = 36
    left = 7, right ===>

    '''
    def maxArea(self, heights: List[int]) -> int:
        maxValue = 0
        left, right =  0, len(heights) -1
        while left <= right:
            maxValue = max(maxValue,min(heights[left],heights[right]) * (right - left))
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1
        
        return maxValue

        